File: Algorithms/109. Convert Sorted List to Binary Search Tree.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
//     pub fn sorted_list_to_bst(head: Option<Box<ListNode>>) -> Option<Rc<RefCell<TreeNode>>> {
//         if head == None {
//             return None;
//         }
//         let mut head = head;
//         let mut mid: Option<Box<ListNode>> = Self::findMiddleElement(&head);
//         let mut node: TreeNode = TreeNode::new(mid.as_ref().unwrap().val);
        
//         if head == mid {
//             return Some(Rc::new(RefCell::new(node)));
//         }
        
//         node.left = Self::sorted_list_to_bst(head);
//         node.right = Self::sorted_list_to_bst(mid.unwrap().next);
//         Some(Rc::new(RefCell::new(node)))
//     }
    
//     pub fn findMiddleElement(head: &Option<Box<ListNode>>) -> Option<Box<ListNode>> {
//         let mut prev: Option<Box<ListNode>> = None;
//         let mut slow: Option<Box<ListNode>> = *head;
//         let mut fast: Option<Box<ListNode>> = *head;
//         while fast != None && fast.as_ref().unwrap().next != None {
//             prev = slow;
//             slow = slow.unwrap().next;
//             fast = fast.unwrap().next.unwrap().next;
//         }
//         if prev != None {
//             prev.unwrap().next = None;
//         }
//         slow
//     }
    pub fn sorted_list_to_bst(head: Option<Box<ListNode>>) -> Option<Rc<RefCell<TreeNode>>> {
        let mut ptr = &head;
        let mut cnt = 0;
        while let Some(boxed) = ptr {
            cnt += 1;
            ptr = &boxed.next
        }
        let mut ptr = &head;
        Self::_list_to_bst(&mut ptr, cnt)
    }

    fn _list_to_bst(ptr: &mut &Option<Box<ListNode>>, len: usize) -> Option<Rc<RefCell<TreeNode>>> {
        if len == 0 {
            return None;
        }
        let left = Self::_list_to_bst(ptr, len / 2);
        let boxed = ptr.as_ref().unwrap();
        let mut node = TreeNode::new(boxed.val);
        node.left = left;
        *ptr = &boxed.next;
        node.right = Self::_list_to_bst(ptr, len - len / 2 - 1);
        Some(Rc::new(RefCell::new(node)))
    }

}

File: Algorithms/109. Convert Sorted List to Binary Search Tree.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function sortedListToBST(head: ListNode | null): TreeNode | null {
  const findMiddleNode = (
    left: ListNode | null,
    right: ListNode | null
  ): ListNode => {
    let fast = left;
    let slow = left;
    while (fast !== right && fast.next !== right) {
      fast = fast.next.next;
      slow = slow.next;
    }
    return slow;
  };

  const buildTree = (left: ListNode | null, right: ListNode | null) => {
    if (left === right) {
      return null;
    }
    const node = findMiddleNode(left, right);
    return new TreeNode(
      node.val,
      buildTree(left, node),
      buildTree(node.next, right)
    );
  };

  return buildTree(head, null);
}
